presto-main/src/main/java/io/prestosql/sql/planner/optimizations/HashGenerationOptimizer.java

```diff
@@ -771,52 +770,56 @@ private PlanWithProperties plan(PlanNode node, HashComputationSet parentPreferen
 
     private static class HashComputationSet
     {
-        private final Set<HashComputation> hashes;
+        private final Map<HashComputation, HashComputation> hashes;
 
         public HashComputationSet()
         {
-            hashes = ImmutableSet.of();
+            hashes = ImmutableMap.of();
         }
 
         public HashComputationSet(Optional<HashComputation> hash)
         {
             requireNonNull(hash, "hash is null");
             if (hash.isPresent()) {
-                this.hashes = ImmutableSet.of(hash.get());
+                this.hashes = ImmutableMap.of(hash.get(), hash.get());
             }
             else {
-                this.hashes = ImmutableSet.of();
+                this.hashes = ImmutableMap.of();
             }
         }
 
-        private HashComputationSet(Iterable<HashComputation> hashes)
+        private HashComputationSet(Map<HashComputation, HashComputation> hashes)
         {
             requireNonNull(hashes, "hashes is null");
-            this.hashes = ImmutableSet.copyOf(hashes);
+            this.hashes = ImmutableMap.copyOf(hashes);
         }
 
         public Set<HashComputation> getHashes()
         {
-            return hashes;
+            return hashes.keySet();
         }
 
         public HashComputationSet pruneSymbols(List<Symbol> symbols)
         {
             Set<Symbol> uniqueSymbols = ImmutableSet.copyOf(symbols);
-            return new HashComputationSet(hashes.stream()
-                    .filter(hash -> hash.canComputeWith(uniqueSymbols))
-                    .collect(toImmutableSet()));
+            return new HashComputationSet(hashes.entrySet().stream()
+                    .filter(hash -> hash.getKey().canComputeWith(uniqueSymbols))
+                    .collect(toImmutableMap(Entry::getKey, Entry::getValue)));
         }
 
         public HashComputationSet translate(Function<Symbol, Optional<Symbol>> translator)
         {
-            Set<HashComputation> newHashes = hashes.stream()
-                    .map(hash -> hash.translate(translator))
-                    .filter(Optional::isPresent)
-                    .map(Optional::get)
-                    .collect(toImmutableSet());
```

Comment:
what was wrong with using `stream` here?

Reply:
As per @martint's suggestion we have moved to `Map<HashComputation, HashComputation>` instead of `Set` and using streams might be a bit complicated (as we need to propagate variables) so removed it